### PR TITLE
Use RBAC for Azure AI Search

### LINKED ISSUE #2444: Azure AI Search should use managed identity by default
See the discussion [here](https://github.com/dotnet/aspire/pull/2070#discussion_r1503262069).

Today we are using a keyvault secret and ApiKey to authenticate to Azure AI Search. We should update this to use managed identity instead.

cc @sebastienros

### PR DESCRIPTION
**Pull request by sebastienros**
Fixes #2444
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2494)

### CODE REVIEW

File: src/Aspire.Hosting.Azure/Bicep/search.bicep [hunk 1]
```diff
@@ -44,7 +44,9 @@ param location string = resourceGroup().location
 
 var resourceToken = uniqueString(resourceGroup().id)
 
-resource search 'Microsoft.Search/searchServices@2020-08-01' = {
+param principalType string = 'ServicePrincipal'
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Bicep/search.bicep:47]: (nit) why aren't principalType and principalId located together in the file?

File: src/Aspire.Hosting.Azure/Bicep/search.bicep [hunk 2]
```diff
@@ -54,21 +56,38 @@ resource search 'Microsoft.Search/searchServices@2020-08-01' = {
...
     partitionCount: partitionCount
     hostingMode: hostingMode
+    disableLocalAuth: true
   }
   tags: tags
 }
 
-var primaryKey = search.listAdminKeys(search.apiVersion).primaryKey
-var searchServiceName = search.name
+// Find list of roles and GUIDs in https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Bicep/search.bicep:64]:
```suggestion
// Find list of roles and GUIDs in https://learn.microsoft.com/azure/role-based-access-control/built-in-roles
```

People get mad when we have locales in our links.

File: src/Aspire.Hosting.Azure/Bicep/search.bicep [hunk 3]
```diff
@@ -54,21 +56,38 @@ resource search 'Microsoft.Search/searchServices@2020-08-01' = {
...
+
+// Search Service Contributor
+var scRoleDefinitionId = subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '7ca78c08-252a-4471-8644-bb5ff32d4ba0')
 
-resource vault 'Microsoft.KeyVault/vaults@2023-07-01' existing = {
-    name: keyVaultName
+// Search Index Data Contributor
+var idRoleDefinitionId2 = subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8ebe5a00-799e-43f5-93ac-243d3dce84a7')
```
Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Bicep/search.bicep:70]: Why is this named `idRoleDefinitionId2`? Why `2`? Why `idRole`?

Comment (eerhardt) [inline src/Aspire.Hosting.Azure/Bicep/search.bicep:70]: (nit) I'd put this right above where it is used. Something like this:

```bicep
// Search Service Contributor
var searchServiceContributorRole = subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '7ca78c08-252a-4471-8644-bb5ff32d4ba0')
resource searchServiceContributorRoleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
  name: guid(search.id, principalId, searchServiceContributorRole)
  scope: search
  properties: {
    principalId: principalId
    principalType: principalType
    roleDefinitionId: searchServiceContributorRole
  }
}

// Search Index Data Contributor
var searchIndexDataContributorRole = subscriptionResourceId('Microsoft.Authorization/roleDefinitions', '8ebe5a00-799e-43f5-93ac-243d3dce84a7')
resource searchIndexDataContributorRoleAssignment 'Microsoft.Authorization/roleAssignments@2022-04-01' = {
  name: guid(search.id, principalId, searchIndexDataContributorRole)
  scope: search
  properties: {
    principalId: principalId
    principalType: principalType
    roleDefinitionId: searchIndexDataContributorRole
  }
}
```

**Comment by sebastienros** [comment]
Also added a missing unit test for the bicep file.

**Review by eerhardt — APPROVED** [review]
LGTM. This is super nice!

**Pull request merged**